src/features/keystrokes/KeystrokesHelper.tsx

```diff
@@ -0,0 +1,137 @@
+import { isEqual } from 'lodash';
+
+import { ExternalLibraryName } from '../../commons/application/types/ExternalTypes';
+import { Input, PlaybackData } from '../sourceRecorder/SourceRecorderTypes';
+
+export const oneHourInMilliSeconds = 3600000;
+const oneByteInBits = 8;
+const oneKbInBytes = 1024;
+const fiveMbInKb = 5 * 1024;
+
+const getLocalStorageSpace = () => {
+  let allStrings = '';
+  for (const key in window.localStorage) {
+    if (window.localStorage.hasOwnProperty(key)) {
+      allStrings += window.localStorage[key];
+    }
+  }
+  return allStrings ? 3 + (allStrings.length * 16) / (oneByteInBits * oneKbInBytes) : 0;
+};
+
+export const hasExceededLocalStorageSpace = () => {
+  return getLocalStorageSpace() > fiveMbInKb;
+};
+
+export const playgroundQuestionId: number = -1;
+
+export const resetPlaygroundLogging = (
+  chapter: number,
+  externalLibrary: ExternalLibraryName,
+  editorValue: string
+) => {
+  const playgroundLogs: string | null = localStorage.getItem('PlaygroundLogs');
+  const playgroundPlayback: PlaybackData = JSON.parse(
+    playgroundLogs ? playgroundLogs : JSON.stringify(defaultPlaybackData)
+  );
+  const newInit = {
+    chapter: chapter,
+    externalLibrary: externalLibrary,
+    editorValue: editorValue
+  };
+  if (!isEqual(playgroundPlayback.init, newInit)) {
+    playgroundPlayback.init = newInit;
+  }
+  playgroundPlayback.inputs = [];
+  localStorage.setItem('PlaygroundLogs', JSON.stringify(playgroundPlayback));
+};
+
+export const savePlaygroundLog = (newInput: Input) => {
+  const playgroundLogs: string | null = localStorage.getItem('PlaygroundLogs');
+  const playgroundPlayback: PlaybackData = JSON.parse(
+    playgroundLogs ? playgroundLogs : JSON.stringify(defaultPlaybackData)
+  );
+  playgroundPlayback.inputs.push(newInput);
+  localStorage.setItem('PlaygroundLogs', JSON.stringify(playgroundPlayback));
+
+  const neww = localStorage.getItem('PlaygroundLogs');
+  const hehexd: PlaybackData = JSON.parse(neww ? neww : JSON.stringify(defaultPlaybackData));
+  console.log(hehexd);
+};
```

Comment:
Which part of the browser prevents them from opening more than 1 playground tab?

What happens when they do so?

What happens when 1 set of logs upload while the user is still on the playground?

Reply:
Also what kind of naming is that.

Reply:
Original lines being reviewed.
```
export const savePlaygroundLog = (newInput: Input) => {
  const playgroundLogs: string | null = localStorage.getItem('PlaygroundLogs');
  const playgroundPlayback: PlaybackData = JSON.parse(
    playgroundLogs ? playgroundLogs : JSON.stringify(defaultPlaybackData)
  );
  playgroundPlayback.inputs.push(newInput);
  localStorage.setItem('PlaygroundLogs', JSON.stringify(playgroundPlayback));

  const neww = localStorage.getItem('PlaygroundLogs');
  const hehexd: PlaybackData = JSON.parse(neww ? neww : JSON.stringify(defaultPlaybackData));
  console.log(hehexd);
};
```

Reply:
O wait I forgot to remove the console.logs. Psps

Reply:
If they open more than one tab, the keystroke will will still get pushed onto our `localStorage` !

Reply:
And if you try to reconstruct the state of both tabs what happens?